From 636767e20cc0615271d0d403c74ee7d21c4a36f7 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Mon, 17 Jun 2019 20:03:49 +0300
Subject: [PATCH] [SW-4352] goya: handle most of plls

Add handling of EMMC, PCI, TPC PLLs in addition to those already handled
in preboot.
Additionally, add the PCI PLL handling for PLDM as till now it was
configured by U-Boot.

Change-Id: I42283afdbc166b3dc516174fc9e2bb319739399b
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/goya/asic.c b/goya/asic.c
index 4a4f510..df00de4 100644
--- a/goya/asic.c
+++ b/goya/asic.c
@@ -26,6 +26,15 @@
 			     HL_PLL_DIV_PLL
 		},
 	}, {
+		.name = "EMMC", .base = mmPSOC_EMMC_PLL_BASE,
+		.nr = 24, .nf = 415, .od = 15, .nb = 207,
+		.div_factor = { 0, 1, -1, -1 },
+		.div_sel = { HL_PLL_PLL_CLK,
+			     HL_PLL_DIV_PLL,
+			     HL_PLL_REF_CLK,
+			     HL_PLL_REF_CLK
+		},
+	}, {
 		.name = "MME", .base = mmPSOC_MME_PLL_BASE,
 		.nr = 0, .nf = 63, .od = 15, .nb = 31,
 		.div_factor = { 0, 1, 2, 8 },
@@ -34,6 +43,24 @@
 			     HL_PLL_DIV_PLL,
 			     HL_PLL_DIV_PLL
 		},
+	}, {
+		.name = "PCI", .base = mmPSOC_PCI_PLL_BASE,
+		.nr = 4, .nf = 343, .od = 3, .nb = 171,
+		.div_factor = { 0, 35, 8, 85 },
+		.div_sel = { HL_PLL_PLL_CLK,
+			     HL_PLL_DIV_PLL,
+			     HL_PLL_DIV_PLL,
+			     HL_PLL_DIV_PLL
+		},
+	}, {
+		.name = "TPC", .base = mmTPC_PLL_BASE,
+		.nr = 0, .nf = 63, .od = 15, .nb = 31,
+		.div_factor = { 0, 1, 1, 8 },
+		.div_sel = { HL_PLL_PLL_CLK,
+			     HL_PLL_DIV_PLL,
+			     HL_PLL_DIV_PLL,
+			     HL_PLL_DIV_PLL
+		},
 	},
 };
 
diff --git a/goya/common.h b/goya/common.h
index 4e2256e..f6e0b2d 100644
--- a/goya/common.h
+++ b/goya/common.h
@@ -7,5 +7,7 @@
 
 #define mmCPU_PLL_BASE		0x4A2000
 #define mmIC_PLL_BASE		0x4A3000
+#define mmPSOC_EMMC_PLL_BASE	0xC70000
 #define mmPSOC_MME_PLL_BASE	0xC71000
 #define mmPSOC_PCI_PLL_BASE	0xC72000
+#define mmTPC_PLL_BASE		0xE01000
diff --git a/goya/pldm.c b/goya/pldm.c
index 24422e8..2cbbc97 100644
--- a/goya/pldm.c
+++ b/goya/pldm.c
@@ -11,16 +11,35 @@
 #include <pll.h>
 #include <uart.h>
 
+#include "common.h"
+
 #define SYS_CLOCK_FREQUENCY	150000ULL
+
+static struct hl_pll_cfg goya_pll_cfg[] = {
+	{
+		.name = "PCI", .base = mmPSOC_PCI_PLL_BASE,
+		.nr = 4, .nf = 343, .od = 3, .nb = 171,
+		.div_factor = { 0, 4, 8, 85 },
+		.div_sel = { HL_PLL_PLL_CLK,
+			     HL_PLL_DIV_PLL,
+			     HL_PLL_DIV_PLL,
+			     HL_PLL_DIV_PLL
+		},
+	},
+};
 
 bool hl_pll_is_configurable(struct hl_pll_cfg *cfg)
 {
-	return false;
+	return cfg->base == mmPSOC_PCI_PLL_BASE;
 }
 
 int hl_pll_get_cfg(struct hl_pll_cfg **hl_pll_cfg)
 {
-	return 0;
+	int pll_num = ARRAY_SIZE(goya_pll_cfg);
+
+	*hl_pll_cfg = goya_pll_cfg;
+
+	return pll_num;
 }
 
 u32 hl_skip_pcie_init(void)
